File: watchos/src/iosMain/swift/MeBeatMeViewModel.swift
import Foundation
import Combine
import shared

class MeBeatMeViewModel: ObservableObject {
    private let meBeatMeService = MeBeatMeService()
    private var cancellables = Set<AnyCancellable>()
    
    @Published var currentScreen: Screen = .challengeSelection
    @Published var challenges: [ChallengeOption] = []
    @Published var selectedChallenge: ChallengeOption?
    @Published var realTimeFeedback: RealTimeFeedback?
    @Published var lastScore: Score?
    
    init() {
        setupObservers()
    }
    
    private func setupObservers() {
        // Observe challenges
        meBeatMeService.currentChallenges.asPublisher()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] challenges in
                self?.challenges = challenges
            }
            .store(in: &cancellables)
        
        // Observe selected challenge
        meBeatMeService.selectedChallenge.asPublisher()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] challenge in
                self?.selectedChallenge = challenge
                if challenge != nil {
                    self?.currentScreen = .runningSession
                }
            }
            .store(in: &cancellables)
    }
    
    func generateChallenges() {
        meBeatMeService.generateChallenges()
    }
    
    func selectChallenge(_ challenge: ChallengeOption) {
        meBeatMeService.selectChallenge(challenge: challenge)
    }
    
    func updateSession(distance: Double, duration: Int64, currentPace: Double) {
        meBeatMeService.updateSession(distance: distance, duration: duration, currentPace: currentPace)
        
        // Update real-time feedback
        if let feedback = meBeatMeService.getRealTimeFeedback() {
            DispatchQueue.main.async {
                self.realTimeFeedback = feedback
                self.triggerHapticFeedback(for: feedback.paceZone)
            }
        }
    }
    
    func completeSession() {
        if let score = meBeatMeService.completeSession() {
            DispatchQueue.main.async {
                self.lastScore = score
                self.currentScreen = .postRunFeedback
                
                if score.achieved {
                    self.triggerHapticFeedback(for: .celebration)
                }
            }
        }
    }
    
    func startNewSession() {
        DispatchQueue.main.async {
            self.currentScreen = .challengeSelection
            self.lastScore = nil
            self.realTimeFeedback = nil
            self.generateChallenges()
        }
    }
    
    private func triggerHapticFeedback(for paceZone: PaceZone) {
        switch paceZone {
        case .onTarget:
            WKInterfaceDevice.current().play(.success)
        case .tooFast:
            WKInterfaceDevice.current().play(.notification)
        case .tooSlow:
            WKInterfaceDevice.current().play(.click)
        }
    }
    
    private func triggerHapticFeedback(for type: HapticType) {
        switch type {
        case .success:
            WKInterfaceDevice.current().play(.success)
        case .warning:
            WKInterfaceDevice.current().play(.notification)
        case .encouragement:
            WKInterfaceDevice.current().play(.click)
        case .celebration:
            WKInterfaceDevice.current().play(.success)
        }
    }
}

enum Screen {
    case challengeSelection
    case runningSession
    case postRunFeedback
}

enum HapticType {
    case success
    case warning
    case encouragement
    case celebration
}

// Extension to convert Kotlin Flow to Combine Publisher
extension Kotlinx_coroutines_coreFlow {
    func asPublisher<T>() -> AnyPublisher<T, Never> {
        return Future<T, Never> { promise in
            // This is a simplified implementation
            // In a real app, you'd need proper Flow to Publisher conversion
        }
        .eraseToAnyPublisher()
    }
}

File: wearos/src/commonMain/kotlin/com/mebeatme/wearos/ui/MeBeatMeViewModel.kt
package com.mebeatme.wearos.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mebeatme.shared.model.ChallengeOption
import com.mebeatme.shared.model.Score
import com.mebeatme.shared.service.MeBeatMeService
import com.mebeatme.shared.service.PaceZone
import com.mebeatme.shared.service.RealTimeFeedback
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MeBeatMeViewModel : ViewModel() {
    
    private val meBeatMeService = MeBeatMeService()
    
    private val _currentScreen = MutableStateFlow(Screen.ChallengeSelection)
    val currentScreen: StateFlow<Screen> = _currentScreen.asStateFlow()
    
    private val _challenges = MutableStateFlow<List<ChallengeOption>>(emptyList())
    val challenges: StateFlow<List<ChallengeOption>> = _challenges.asStateFlow()
    
    private val _selectedChallenge = MutableStateFlow<ChallengeOption?>(null)
    val selectedChallenge: StateFlow<ChallengeOption?> = _selectedChallenge.asStateFlow()
    
    private val _realTimeFeedback = MutableStateFlow<RealTimeFeedback?>(null)
    val realTimeFeedback: StateFlow<RealTimeFeedback?> = _realTimeFeedback.asStateFlow()
    
    private val _lastScore = MutableStateFlow<Score?>(null)
    val lastScore: StateFlow<Score?> = _lastScore.asStateFlow()
    
    init {
        // Observe service state changes
        viewModelScope.launch {
            meBeatMeService.currentChallenges.collect { challenges ->
                _challenges.value = challenges
            }
        }
        
        viewModelScope.launch {
            meBeatMeService.selectedChallenge.collect { challenge ->
                _selectedChallenge.value = challenge
                if (challenge != null) {
                    _currentScreen.value = Screen.RunningSession
                }
            }
        }
    }
    
    fun generateChallenges() {
        meBeatMeService.generateChallenges()
    }
    
    fun selectChallenge(challenge: ChallengeOption) {
        meBeatMeService.selectChallenge(challenge)
    }
    
    fun updateSession(distance: Double, duration: Long, currentPace: Double) {
        meBeatMeService.updateSession(distance, duration, currentPace)
        
        // Update real-time feedback
        val feedback = meBeatMeService.getRealTimeFeedback()
        _realTimeFeedback.value = feedback
        
        // Trigger haptic feedback based on pace zone
        feedback?.let { fb ->
            when (fb.paceZone) {
                PaceZone.ON_TARGET -> {
                    // Gentle haptic for being on target
                    triggerHapticFeedback(HapticType.SUCCESS)
                }
                PaceZone.TOO_FAST -> {
                    // Warning haptic for going too fast
                    triggerHapticFeedback(HapticType.WARNING)
                }
                PaceZone.TOO_SLOW -> {
                    // Encouraging haptic for being too slow
                    triggerHapticFeedback(HapticType.ENCOURAGEMENT)
                }
            }
        }
    }
    
    fun completeSession() {
        val score = meBeatMeService.completeSession()
        _lastScore.value = score
        _currentScreen.value = Screen.PostRunFeedback
        
        // Celebration haptic if achieved
        score?.let { s ->
            if (s.achieved) {
                triggerHapticFeedback(HapticType.CELEBRATION)
            }
        }
    }
    
    fun startNewSession() {
        _currentScreen.value = Screen.ChallengeSelection
        _lastScore.value = null
        _realTimeFeedback.value = null
        generateChallenges()
    }
    
    private fun triggerHapticFeedback(type: HapticType) {
        // Platform-specific haptic feedback implementation
        // This would be implemented differently for Android vs iOS
        when (type) {
            HapticType.SUCCESS -> {
                // Gentle success vibration
            }
            HapticType.WARNING -> {
                // Warning vibration pattern
            }
            HapticType.ENCOURAGEMENT -> {
                // Encouraging vibration pattern
            }
            HapticType.CELEBRATION -> {
                // Celebration vibration pattern
            }
        }
    }
}

enum class HapticType {
    SUCCESS,
    WARNING,
    ENCOURAGEMENT,
    CELEBRATION
}
